File: leetcode/Binary Search Tree to Greater Sum Tree/2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func bstToGst(root *TreeNode) *TreeNode {
	stack := []*TreeNode{root}
	nodes := []*TreeNode{}
	for len(stack) > 0 {
		curr := stack[0]
		stack = stack[1:]
		if curr == nil {
			continue
		}
		nodes = append(nodes, curr)
		stack = append(stack, curr.Left, curr.Right)
	}

	sort.SliceStable(nodes, func(i, j int) bool {
		return nodes[i].Val < nodes[j].Val
	})

	for i := len(nodes) - 2; i >= 0; i-- {
		nodes[i].Val += nodes[i+1].Val
	}
	return root
}

File: leetcode/Lowest Common Ancestor of a Binary Search Tree/2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        if (root == null) return null

        val pair = getMinMax(p!!.`val`, q!!.`val`)
        val min = pair.first
        val max = pair.second
        val rootVal = root?.`val` ?: 0

        if (rootVal in min..max) return root
        if (rootVal > max) return lowestCommonAncestor(root.left, p, q)
        return lowestCommonAncestor(root.right, p, q)
    }

    private fun getMinMax(a: Int, b: Int): Pair<Int, Int> =
        if (a < b) Pair(a, b) else Pair(b, a)

}